flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/ParameterizedPredicate.java

```diff
@@ -52,8 +69,26 @@ public void setPredicate(String predicate) {
     }
 
     public ParameterizedPredicate combine(String operator, ParameterizedPredicate that) {
+        int paramIndex = String.format("(%s %s ", this.predicate, operator).length();
+        if (!that.indexesOfPredicatePlaceHolders.isEmpty()) {
+            paramIndex = paramIndex + that.indexesOfPredicatePlaceHolders.get(0);
+        }
+
         this.predicate = String.format("(%s %s %s)", this.predicate, operator, that.predicate);
         this.parameters = ArrayUtils.addAll(this.parameters, that.parameters);
+
+        for (int i = 0; i < this.indexesOfPredicatePlaceHolders.size(); i++) {
+            // increment all the existing indexes to account for the new additional first begin
+            // bracket
+            this.indexesOfPredicatePlaceHolders.set(
+                    i, this.indexesOfPredicatePlaceHolders.get(i) + 1);
+        }
+        if (that.predicate.equals(
+                        JdbcFilterPushdownPreparedStatementVisitor.PUSHDOWN_PREDICATE_PLACEHOLDER)
+                || (!that.indexesOfPredicatePlaceHolders.isEmpty())) {
+            // add index if that is a placeholder or has a placeholder.
+            this.indexesOfPredicatePlaceHolders.add(paramIndex);
+        }
```

Comment:
This will not work for cases when constant on the left side e.g.
```sql
    @Test
    public void test123() {
        util.verifyExecPlan(
                "SELECT * FROM a left join d FOR SYSTEM_TIME AS OF a.proctime on (50 = d.age OR 1 = d.type)  and a.ip = d.ip");
    }
```
this test will fail with 
```
Caused by: java.lang.IndexOutOfBoundsException: Index: 0, Size: 0
	at java.util.ArrayList.rangeCheck(ArrayList.java:659)
	at java.util.ArrayList.get(ArrayList.java:435)
	at org.apache.flink.connector.jdbc.table.JdbcDynamicTableSource.getResolvedConditions(JdbcDynamicTableSource.java:157)
	at org.apache.flink.connector.jdbc.table.JdbcDynamicTableSource.getLookupRuntimeProvider(JdbcDynamicTableSource.java:120)
	at org.apache.flink.table.planner.plan.utils.LookupJoinUtil.createLookupRuntimeProvider(LookupJoinUtil.java:626)
	at org.apache.flink.table.planner.plan.utils.LookupJoinUtil.isAsyncLookup(LookupJoinUtil.java:413)
```

The method itself makes some things which could be omitted, I would suggest to replace it with 
```suggestion
        this.predicate = String.format("(%s %s %s)", this.predicate, operator, that.predicate);
        this.parameters = ArrayUtils.addAll(this.parameters, that.parameters);

        this.indexesOfPredicatePlaceHolders.add("?".equals(this.predicate) ? 0 : this.predicate.length() - 1);
```
which is simpler and does not fail for the mentioned test
WDYT?

Reply:
Also I tend to think that approach with indexes will not work because between finding indexes and building `resolvedConditions` there is an optimizer which could change SQL a bit
example of query where such change happens
```sql
"SELECT * FROM a left join d FOR SYSTEM_TIME AS OF a.proctime on ((d.age = 50 AND d.type = 0) OR (d.type = 1 AND d.age =40))  and a.ip = d.ip"
```

Reply:
@snuyanzin I have amended the code locally to work with `?` on the left. Also queries of the type 
`SELECT * FROM a left join d FOR SYSTEM_TIME AS OF a.proctime on ((d.age = 50 AND d.type = 0) OR (d.type = 1 AND d.age =40))  and a.ip = d.ip;` are working for me locally . 

I intend to put up the latest fix once I have done more testing. But if you think the index design has a more fundamental flaw, then we will need to rethink. 

I appreciate you drawing attention to the queries that are not working. The latest fix stashes the indexes in the visitor if the indexes are correct then we can replace parameters only where we know we should.